pipeline.lisp

```diff
@@ -15,28 +15,64 @@
    (%names :initarg :names :initform (make-hash-table :test 'equal) :accessor names))
   (:documentation "Base pipeline object that is necessary to orchestrate piping systems."))
 
+(define-condition unknown-name-error ()
+  ((%name :initarg :name :reader name)
+   (%pipeline :initarg :pipeline :reader pipeline))
+  (:documentation "Signalled if a name which was not previously set in the pipeline is looked up")
+  (:report (lambda (c stream)
+             (format stream "Name ~A is not known in pipeline ~A" (name c) (pipeline c)))))
+
+(defgeneric resolve-place (pipeline place &key ignore-unknown)
+  (:documentation "Resolve PLACE to a path (ie. list of integers). If PLACE is already a
+  path, return it unchanged. If it's a name (as set by SET-NAME), return its corresponding
+  path. If the name is not known, signal UNKNOWN-NAME-ERROR, unless IGNORE-UNKNOWN is
+  non-nil, in which case NIL is returned.
+
+  Returns a secondary value which is T if the place was successfully resolved, and NIL if
+  it was an unknown name and INGORE-UNKNOWN was specified.")
+  (:method ((pipeline pipeline) (place list) &key &allow-other-keys)
+    (values place t))
+  ;; NIL is a valid path with no segments, pointing to the entire pipeline
+  (:method ((pipeline pipeline) (place null) &key &allow-other-keys)
+    (values place t))
+  (:method ((pipeline pipeline) (place symbol) &key ignore-unknown)
+    (multiple-value-bind (path found)
+        (gethash place (names pipeline))
```

Comment:
It isn't valid to store a NIL path as a named path, so using the `found` complication here seems unnecessary to me.

Reply:
OK, if `NIL` is specifically invalid, then I can certainly simplify this.